### ISSUE TITLE
RFC: controlled vs uncontrolled inputs

### ISSUE THREAD

**Issue by SiAdcock** [guardian/source#360]
## Uncontrolled

Currently Source input components such as Checkbox and Radio use an [uncontrolled](https://reactjs.org/docs/uncontrolled-components.html) approach to state management. The state is stored in the DOM rather than within the React state. 

It is not necessary to explicitly set a default value, nor have that value stored in React state.

Consumers of these components can access the state from the DOM in a `onChange` handler. e.g

```jsx
<Checkbox
  label="Check me"
  onChange={e => {
    respondToChange(e.target.checked);
  }}
/>
```

In the uncontrolled approach, a consumer could also access the state using a `ref`. This is not currently possible as ref forwarding is not enabled on any Source components (see #294).

## Controlled

As a shift in approach we could make Source components [controlled](https://reactjs.org/docs/forms.html#controlled-components). This makes React the single source of truth for component state. It forces the consumer to specify an initial state as well as an onChange handler to access the state.

```jsx
const {isChecked, setChecked} = useState(false)

return (
  <Checkbox
    label="Check me"
    checked={isChecked}
    onChange={e => {
      setChecked(e.target.checked);
    }}
  />
)
```

React recommend this approach, and it also unlocks more powerful features such as [immediate field validation](https://goshakkk.name/instant-form-fields-validation-react/), or forms that allow you to [dynamically add or delete form fields](https://goshakkk.name/array-form-inputs/).

## Request for comments

I'm leaning towards making all our form components controlled, mainly for more powerful features. It means we can provide stricter type enforcement, by insisting on consumers providing a default value and an onChange handler. It helps us work around tricky bugs that surface when you're not sure if React or the DOM is the source of truth for state (cc @twrichards @rBangay)

Does anyone have any opinions, concerns or requests for clarification on this? I'm super interested to hear from you, and especially anyone with experience building React forms, either simple or complex.

**Comment by jonathonherbert** [comment]
I'd agree that controlled inputs would be best.

The exceptions to this rule that I've seen have generally been performance related – for example, when the cost of passing the output of a rich text editor through the state management loop has had an impact on the user experience.

If that's not a concern here, then controlled inputs sound 👌

**Comment by jeanlauliac** [comment]
From my experience building complex forms at Facebook, go 100% with controlled components. With uncontrolled components, it's not long before you get duplicated state, the state of the DOM and the state you inevitably need to process the data; and duplicated state soon leads to out-of-sync values.

For rich text editors, use pre-existing solutions like https://draftjs.org/. These are very hard to get right, and Draft.js already provides an optimized controlled React component.

**Comment by JamieB-gu** [comment]
I would agree with the comments above. I think using controlled components avoids the possibility of state synchronisation bugs.

**Comment by twrichards** [comment]
Thank you @SiAdcock for following up so diligently, I had totally forgotten about this.

As the originator of the suggestion that `checked` be mandatory to encourage us consumers of _Source_ not to make mistakes/bugs (it was quite a hard bug to track down), it would be **great** to see these become 'controlled'. 😄

**Comment by SiAdcock** [comment]
## Update

In v1.9.0, we're adding controlled component support to every form component in Source. This essentially means fixing animation bugs, ensuring that React is the single source of truth for state (where possibel) and updating the documentation to recommend using controlled components. 

However, we're stopping short of insisting that all components must be controlled by React because this has some drawbacks. Notably they require a fair bit of boilerplate, which is unnecessary for server-rendered forms.

We may revists this in the future if there's a compelling case to remove support for uncontrolled components.

Thanks for all your feedback.

**Comment by SiAdcock** [comment]
Addressed in #432, #434, #436, #437, #438
